src/model/utils/selection-post-fixer.js

```diff
@@ -146,27 +146,87 @@ function tryFixingCollapsedRange( range, schema ) {
 	return new Range( fixedPosition );
 }
 
-// Tries to fix a expanded range that overlaps limit nodes.
+// Tries to fix an expanded range.
 //
 // @param {module:engine/model/range~Range} range Expanded range to fix.
 // @param {module:engine/model/schema~Schema} schema
 // @returns {module:engine/model/range~Range|null} Returns fixed range or null if range is valid.
-function tryFixingNonCollpasedRage( range, schema ) {
-	// No need to check flat ranges as they will not cross node boundary.
-	if ( range.isFlat ) {
+function tryFixingNonCollapsedRage( range, schema ) {
+	const start = range.start;
+	const end = range.end;
+
+	// Flat range on the same text node is always valid - no need to fix:
```

Comment:
I think that this comment isn't entirely correct – I mean the "same" part. What about:

```html
<p>f[oo<inlineWidget/>ba]r</p>
```

Also, what about:

```html
<p>[foo]</p>
```

In this case, neither the start's nor end's `textNode` getter will return something.

And there's one more case which is missed here:

```html
<p>[<inlineWidget/>foo<inlineWidget/>]</p>
```

You might've had a different goal by adding this condition – filtering some cases which might be broken by the following logic of this function. But, if the logic of such a check isn't "complete" in a bigger picture, a wider context of the entire algorithm, then it may be misleading. So, instead of patching this algorithm for a hidden purpose, let's thing wider.

E.g. what kind of non-collapsed selections don't need to be fixed? Those which meet the following criteria (all):

* `schema.getLimitElement( range.start ) == schema.getLimitElement( range.end )` – it means the selection starts and ends in the same limit element, so it does not cross any limit element boundary. If a range violates this rule, it needs to be extended to cover the lowest common limit element of both its ends (`schem.getLimitElement( range )`).

   Note: Right now, `schema.getLimitElement()` works only with entire selections, but it's not hard to "downgrade" it to support single ranges (returns LCA like for the entire selection) and single positions (returns a lowest limit element in which this position is contained) too.
* `schema.checkChild( range.start, '$text' ) && schema.checkChild( range.end, '$text' )` – both its ends are in a place where text is allowed. They don't need to be placed in/next to a text (`<p>[<inlineWidget/>...`) but it's enough that text is allowed there.

   Note: A range containing a block widget will not be rooted in a place where text is allowed, so it will pass this check. However, it doesn't mean it's incorrect – I was describing a range which is certainly correct (i.e. a possible early return check), not an algorithm for deciding whether a range is correct or not.

Reply:
By structuring the algorithm as I described above, the rules by which it works will be easier to read. Right now, it consists of "let's try fixing this, let's try that" which is harder for me to digest.